// api/dist/src/notifications/enums/notification-type.enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NotificationType = void 0;
var NotificationType;
(function (NotificationType) {
    NotificationType["WELCOME"] = "WELCOME";
    NotificationType["APPOINTMENT_REMINDER"] = "APPOINTMENT_REMINDER";
    NotificationType["APPOINTMENT_CONFIRMATION"] = "APPOINTMENT_CONFIRMATION";
    NotificationType["APPOINTMENT_CANCELLATION"] = "APPOINTMENT_CANCELLATION";
    NotificationType["TEST_RESULTS"] = "TEST_RESULTS";
    NotificationType["PRESCRIPTION_READY"] = "PRESCRIPTION_READY";
    NotificationType["BILLING_REMINDER"] = "BILLING_REMINDER";
    NotificationType["PAYMENT_CONFIRMATION"] = "PAYMENT_CONFIRMATION";
    NotificationType["NEW_REFERRAL"] = "NEW_REFERRAL";
    NotificationType["NEW_TEST_ORDER"] = "NEW_TEST_ORDER";
    NotificationType["ABNORMAL_TEST_RESULT"] = "ABNORMAL_TEST_RESULT";
    NotificationType["STAFF_SCHEDULE_UPDATE"] = "STAFF_SCHEDULE_UPDATE";
    NotificationType["PASSWORD_RESET"] = "PASSWORD_RESET";
    NotificationType["EMAIL_VERIFICATION"] = "EMAIL_VERIFICATION";
    NotificationType["PHONE_VERIFICATION"] = "PHONE_VERIFICATION";
    NotificationType["ACCOUNT_ACTIVATION"] = "ACCOUNT_ACTIVATION";
    NotificationType["SECURITY_ALERT"] = "SECURITY_ALERT";
    NotificationType["NEW_MESSAGE"] = "NEW_MESSAGE";
    NotificationType["ANNOUNCEMENT"] = "ANNOUNCEMENT";
    NotificationType["FEEDBACK_REQUEST"] = "FEEDBACK_REQUEST";
    NotificationType["SURVEY"] = "SURVEY";
})(NotificationType || (exports.NotificationType = NotificationType = {}));
