// api/dist/src/patients/interceptors/patient-audit.interceptor.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PatientAuditInterceptor = void 0;
const common_1 = require("@nestjs/common");
const operators_1 = require("rxjs/operators");
const audit_service_1 = require("../../audit/audit.service");
let PatientAuditInterceptor = class PatientAuditInterceptor {
    auditService;
    constructor(auditService) {
        this.auditService = auditService;
    }
    intercept(context, next) {
        const httpContext = context.switchToHttp();
        const request = httpContext.getRequest();
        const response = httpContext.getResponse();
        const { method, url, body, params, user, headers } = request;
        const tenantId = headers['x-tenant-id'];
        const userId = user?.id || 'system';
        if (method === 'GET') {
            return next.handle();
        }
        const action = this.getActionFromMethod(method);
        const resourceId = params?.id || body?.id;
        let oldValues = null;
        let newValues = null;
        if (['POST', 'PUT', 'PATCH'].includes(method)) {
            newValues = { ...body };
            delete newValues.password;
            delete newValues.token;
        }
        return next.handle().pipe((0, operators_1.tap)(async () => {
            try {
                await this.auditService.log({
                    tenantId,
                    userId,
                    action,
                    resource: 'Patient',
                    resourceId,
                    oldValues,
                    newValues,
                    ipAddress: request.ip,
                    userAgent: headers['user-agent'] || '',
                    timestamp: new Date(),
                    statusCode: response.statusCode,
                    metadata: {
                        url,
                        method,
                        params,
                    },
                });
            }
            catch (error) {
                console.error('Failed to log patient audit:', error);
            }
        }));
    }
    getActionFromMethod(method) {
        switch (method) {
            case 'POST':
                return 'CREATE';
            case 'PATCH':
            case 'PUT':
                return 'UPDATE';
            case 'DELETE':
                return 'DELETE';
            default:
                return 'UNKNOWN';
        }
    }
};
exports.PatientAuditInterceptor = PatientAuditInterceptor;
exports.PatientAuditInterceptor = PatientAuditInterceptor = __decorate([
    (0, common_1.Injectable)(),
    __metadata("design:paramtypes", [audit_service_1.AuditService])
], PatientAuditInterceptor);
