// api/dist/src/opd-management/repositories/order.repository.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderRepository = void 0;
const typeorm_1 = require("typeorm");
const order_entity_1 = require("../entities/order.entity");
let OrderRepository = class OrderRepository extends typeorm_1.Repository {
    async findByVisitId(visitId) {
        return this.find({
            where: { visitId },
            relations: ['provider', 'items'],
            order: { orderDate: 'DESC' },
        });
    }
    async findByPatientId(patientId, status) {
        const where = { patientId };
        if (status)
            where.status = status;
        return this.find({
            where,
            relations: ['provider', 'visit'],
            order: { orderDate: 'DESC' },
        });
    }
    async findPendingOrders(providerId) {
        const where = {
            status: (0, typeorm_1.In)([order_entity_1.OrderStatus.REQUESTED, order_entity_1.OrderStatus.IN_PROGRESS])
        };
        if (providerId)
            where.providerId = providerId;
        return this.find({
            where,
            relations: ['patient', 'provider', 'visit'],
            order: {
                priority: 'DESC',
                orderDate: 'ASC'
            },
        });
    }
    async findOrdersByType(type, status, startDate, endDate) {
        const where = { type };
        if (status)
            where.status = status;
        if (startDate && endDate) {
            where.orderDate = (0, typeorm_1.Between)(startDate, endDate);
        }
        return this.find({
            where,
            relations: ['patient', 'provider', 'visit'],
            order: { orderDate: 'DESC' },
        });
    }
    async getOrderStats(providerId) {
        const where = {};
        if (providerId)
            where.providerId = providerId;
        const [total, completed, pending, inProgress, cancelled] = await Promise.all([
            this.count({ where }),
            this.count({ where: { ...where, status: order_entity_1.OrderStatus.COMPLETED } }),
            this.count({ where: { ...where, status: order_entity_1.OrderStatus.REQUESTED } }),
            this.count({ where: { ...where, status: order_entity_1.OrderStatus.IN_PROGRESS } }),
            this.count({ where: { ...where, status: order_entity_1.OrderStatus.CANCELLED } }),
        ]);
        const typeResults = await this.createQueryBuilder('order')
            .select('order.type', 'type')
            .addSelect('COUNT(order.id)', 'count')
            .where(providerId ? 'order.providerId = :providerId' : '1=1', { providerId })
            .groupBy('order.type')
            .getRawMany();
        const byType = typeResults.reduce((acc, { type, count }) => ({
            ...acc,
            [type]: parseInt(count, 10),
        }), {});
        return {
            total,
            completed,
            pending,
            inProgress,
            cancelled,
            byType,
        };
    }
    async findOrdersNeedingAttention(providerId) {
        const query = this.createQueryBuilder('order')
            .where('order.status IN (:...statuses)', {
            statuses: [order_entity_1.OrderStatus.REQUESTED, order_entity_1.OrderStatus.IN_PROGRESS],
        })
            .andWhere('order.dueDate <= :today', { today: new Date() })
            .orderBy('order.priority', 'DESC')
            .addOrderBy('order.dueDate', 'ASC');
        if (providerId) {
            query.andWhere('order.providerId = :providerId', { providerId });
        }
        return query
            .leftJoinAndSelect('order.patient', 'patient')
            .leftJoinAndSelect('order.provider', 'provider')
            .getMany();
    }
    async findOrdersByItem(itemName, type) {
        const query = this.createQueryBuilder('order')
            .innerJoin('order.items', 'item')
            .where('LOWER(item.name) LIKE LOWER(:name)', { name: `%${itemName}%` });
        if (type) {
            query.andWhere('order.type = :type', { type });
        }
        return query
            .leftJoinAndSelect('order.patient', 'patient')
            .leftJoinAndSelect('order.provider', 'provider')
            .orderBy('order.orderDate', 'DESC')
            .getMany();
    }
};
exports.OrderRepository = OrderRepository;
exports.OrderRepository = OrderRepository = __decorate([
    (0, typeorm_1.EntityRepository)(order_entity_1.Order)
], OrderRepository);
