// api/dist/src/ipd-management/enums/anesthesia-type.enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnesthesiaType = void 0;
var AnesthesiaType;
(function (AnesthesiaType) {
    AnesthesiaType["GENERAL"] = "GENERAL";
    AnesthesiaType["INHALATION"] = "INHALATION";
    AnesthesiaType["INTRAVENOUS"] = "INTRAVENOUS";
    AnesthesiaType["BALANCED"] = "BALANCED";
    AnesthesiaType["REGIONAL"] = "REGIONAL";
    AnesthesiaType["SPINAL"] = "SPINAL";
    AnesthesiaType["EPIDURAL"] = "EPIDURAL";
    AnesthesiaType["CAUDAL"] = "CAUDAL";
    AnesthesiaType["NERVE_BLOCK"] = "NERVE_BLOCK";
    AnesthesiaType["LOCAL"] = "LOCAL";
    AnesthesiaType["TOPICAL"] = "TOPICAL";
    AnesthesiaType["INFILTRATION"] = "INFILTRATION";
    AnesthesiaType["CONSCIOUS_SEDATION"] = "CONSCIOUS_SEDATION";
    AnesthesiaType["MODERATE_SEDATION"] = "MODERATE_SEDATION";
    AnesthesiaType["DEEP_SEDATION"] = "DEEP_SEDATION";
    AnesthesiaType["PEDIATRIC"] = "PEDIATRIC";
    AnesthesiaType["OBSTETRIC"] = "OBSTETRIC";
    AnesthesiaType["CARDIAC"] = "CARDIAC";
    AnesthesiaType["MONITORED_ANESTHESIA_CARE"] = "MONITORED_ANESTHESIA_CARE";
    AnesthesiaType["TIVA"] = "TIVA";
    AnesthesiaType["NONE"] = "NONE";
    AnesthesiaType["OTHER"] = "OTHER";
})(AnesthesiaType || (exports.AnesthesiaType = AnesthesiaType = {}));
