// api/dist/src/ipd-management/enums/bed-transfer-type.enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BedTransferType = void 0;
var BedTransferType;
(function (BedTransferType) {
    BedTransferType["CLINICAL_DETERIORATION"] = "CLINICAL_DETERIORATION";
    BedTransferType["CLINICAL_IMPROVEMENT"] = "CLINICAL_IMPROVEMENT";
    BedTransferType["SPECIALIZED_CARE_NEEDED"] = "SPECIALIZED_CARE_NEEDED";
    BedTransferType["BED_AVAILABILITY"] = "BED_AVAILABILITY";
    BedTransferType["WARD_REORGANIZATION"] = "WARD_REORGANIZATION";
    BedTransferType["PATIENT_REQUEST"] = "PATIENT_REQUEST";
    BedTransferType["ISOLATION_REQUIRED"] = "ISOLATION_REQUIRED";
    BedTransferType["INFECTION_CONTROL"] = "INFECTION_CONTROL";
    BedTransferType["BEHAVIORAL_ISSUES"] = "BEHAVIORAL_ISSUES";
    BedTransferType["TRANSFER_TO_HIGHER_CARE"] = "TRANSFER_TO_HIGHER_CARE";
    BedTransferType["TRANSFER_TO_LOWER_CARE"] = "TRANSFER_TO_LOWER_CARE";
    BedTransferType["HOSPITAL_TRANSFER"] = "HOSPITAL_TRANSFER";
    BedTransferType["MAINTENANCE"] = "MAINTENANCE";
    BedTransferType["RENOVATION"] = "RENOVATION";
    BedTransferType["OTHER"] = "OTHER";
})(BedTransferType || (exports.BedTransferType = BedTransferType = {}));
