// api/dist/src/ipd-management/enums/consent-type.enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConsentType = void 0;
var ConsentType;
(function (ConsentType) {
    ConsentType["INFORMED_CONSENT"] = "INFORMED_CONSENT";
    ConsentType["SURGICAL_CONSENT"] = "SURGICAL_CONSENT";
    ConsentType["ANESTHESIA_CONSENT"] = "ANESTHESIA_CONSENT";
    ConsentType["BLOOD_TRANSFUSION"] = "BLOOD_TRANSFUSION";
    ConsentType["HIGH_RISK_PROCEDURE"] = "HIGH_RISK_PROCEDURE";
    ConsentType["MEDICAL_TREATMENT"] = "MEDICAL_TREATMENT";
    ConsentType["PSYCHIATRIC_TREATMENT"] = "PSYCHIATRIC_TREATMENT";
    ConsentType["CHEMOTHERAPY"] = "CHEMOTHERAPY";
    ConsentType["RADIATION_THERAPY"] = "RADIATION_THERAPY";
    ConsentType["PRIVACY_CONSENT"] = "PRIVACY_CONSENT";
    ConsentType["RELEASE_OF_INFORMATION"] = "RELEASE_OF_INFORMATION";
    ConsentType["PHOTO_VIDEO_CONSENT"] = "PHOTO_VIDEO_CONSENT";
    ConsentType["RESEARCH_PARTICIPATION"] = "RESEARCH_PARTICIPATION";
    ConsentType["MINOR_TREATMENT"] = "MINOR_TREATMENT";
    ConsentType["EMERGENCY_TREATMENT"] = "EMERGENCY_TREATMENT";
    ConsentType["DO_NOT_RESUSCITATE"] = "DO_NOT_RESUSCITATE";
    ConsentType["ADVANCE_DIRECTIVE"] = "ADVANCE_DIRECTIVE";
    ConsentType["OTHER"] = "OTHER";
})(ConsentType || (exports.ConsentType = ConsentType = {}));
