// api/dist/src/ipd-management/enums/medication-frequency.enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MedicationFrequency = void 0;
var MedicationFrequency;
(function (MedicationFrequency) {
    MedicationFrequency["ONCE_DAILY"] = "ONCE_DAILY";
    MedicationFrequency["TWICE_DAILY"] = "TWICE_DAILY";
    MedicationFrequency["THREE_TIMES_DAILY"] = "THREE_TIMES_DAILY";
    MedicationFrequency["FOUR_TIMES_DAILY"] = "FOUR_TIMES_DAILY";
    MedicationFrequency["EVERY_6_HOURS"] = "EVERY_6_HOURS";
    MedicationFrequency["EVERY_8_HOURS"] = "EVERY_8_HOURS";
    MedicationFrequency["EVERY_12_HOURS"] = "EVERY_12_HOURS";
    MedicationFrequency["EVERY_24_HOURS"] = "EVERY_24_HOURS";
    MedicationFrequency["AS_NEEDED"] = "AS_NEEDED";
    MedicationFrequency["EVERY_4_HOURS_AS_NEEDED"] = "EVERY_4_HOURS_AS_NEEDED";
    MedicationFrequency["EVERY_6_HOURS_AS_NEEDED"] = "EVERY_6_HOURS_AS_NEEDED";
    MedicationFrequency["ONCE_WEEKLY"] = "ONCE_WEEKLY";
    MedicationFrequency["TWICE_WEEKLY"] = "TWICE_WEEKLY";
    MedicationFrequency["ONCE_MONTHLY"] = "ONCE_MONTHLY";
    MedicationFrequency["BEFORE_MEALS"] = "BEFORE_MEALS";
    MedicationFrequency["AFTER_MEALS"] = "AFTER_MEALS";
    MedicationFrequency["AT_BEDTIME"] = "AT_BEDTIME";
    MedicationFrequency["IN_THE_MORNING"] = "IN_THE_MORNING";
    MedicationFrequency["IN_THE_EVENING"] = "IN_THE_EVENING";
    MedicationFrequency["CONTINUOUS"] = "CONTINUOUS";
    MedicationFrequency["CONTINUOUS_INFUSION"] = "CONTINUOUS_INFUSION";
    MedicationFrequency["EVERY_OTHER_DAY"] = "EVERY_OTHER_DAY";
    MedicationFrequency["OTHER"] = "OTHER";
})(MedicationFrequency || (exports.MedicationFrequency = MedicationFrequency = {}));
