// api/dist/src/ipd-management/enums/medication-route.enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MedicationRoute = void 0;
var MedicationRoute;
(function (MedicationRoute) {
    MedicationRoute["ORAL"] = "ORAL";
    MedicationRoute["SUBLINGUAL"] = "SUBLINGUAL";
    MedicationRoute["BUCCAL"] = "BUCCAL";
    MedicationRoute["TOPICAL"] = "TOPICAL";
    MedicationRoute["TRANSDERMAL"] = "TRANSDERMAL";
    MedicationRoute["OCULAR"] = "OCULAR";
    MedicationRoute["OTIC"] = "OTIC";
    MedicationRoute["NASAL"] = "NASAL";
    MedicationRoute["INHALATION"] = "INHALATION";
    MedicationRoute["INTRAVENOUS"] = "INTRAVENOUS";
    MedicationRoute["INTRAMUSCULAR"] = "INTRAMUSCULAR";
    MedicationRoute["SUBCUTANEOUS"] = "SUBCUTANEOUS";
    MedicationRoute["INTRADERMAL"] = "INTRADERMAL";
    MedicationRoute["INTRATHECAL"] = "INTRATHECAL";
    MedicationRoute["EPIDURAL"] = "EPIDURAL";
    MedicationRoute["ENTERAL"] = "ENTERAL";
    MedicationRoute["NASOGASTRIC"] = "NASOGASTRIC";
    MedicationRoute["PERCUTANEOUS_ENDOSCOPIC_GASTROSTOMY"] = "PERCUTANEOUS_ENDOSCOPIC_GASTROSTOMY";
    MedicationRoute["RECTAL"] = "RECTAL";
    MedicationRoute["VAGINAL"] = "VAGINAL";
    MedicationRoute["INTRAOSSEOUS"] = "INTRAOSSEOUS";
    MedicationRoute["INTRAPERITONEAL"] = "INTRAPERITONEAL";
    MedicationRoute["INTRACARDIAC"] = "INTRACARDIAC";
    MedicationRoute["INTRATHECAL"] = "INTRATHECAL";
    MedicationRoute["OTHER"] = "OTHER";
})(MedicationRoute || (exports.MedicationRoute = MedicationRoute = {}));
