// api/dist/src/integrations/services/hl7.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var Hl7Service_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Hl7Service = void 0;
const common_1 = require("@nestjs/common");
const prisma_service_1 = require("../../prisma/prisma.service");
const audit_service_1 = require("../../audit/audit.service");
let Hl7Service = Hl7Service_1 = class Hl7Service {
    prisma;
    auditService;
    logger = new common_1.Logger(Hl7Service_1.name);
    constructor(prisma, auditService) {
        this.prisma = prisma;
        this.auditService = auditService;
    }
    async processHl7Message(hl7Dto, user) {
        this.logger.log(`Processing HL7 message for tenant: ${user.tenantId}`);
        const { messageType, messageContent, sourceSystem } = hl7Dto;
        const parsedMessage = this.parseHl7Message(messageContent);
        let result;
        switch (messageType) {
            case 'ADT_A01':
                result = await this.processAdmitMessage(parsedMessage, user);
                break;
            case 'ADT_A03':
                result = await this.processDischargeMessage(parsedMessage, user);
                break;
            case 'ORU_R01':
                result = await this.processObservationMessage(parsedMessage, user);
                break;
            case 'ORM_O01':
                result = await this.processOrderMessage(parsedMessage, user);
                break;
            default:
                throw new common_1.BadRequestException(`Unsupported HL7 message type: ${messageType}`);
        }
        const hl7Message = await this.prisma.hl7Message.create({
            data: {
                tenantId: user.tenantId,
                messageType,
                messageContent,
                sourceSystem,
                processedAt: new Date(),
                status: 'PROCESSED',
                processedBy: user.id,
            },
        });
        await this.auditService.logActivity({
            action: 'HL7_MESSAGE_PROCESSED',
            entityType: 'HL7_MESSAGE',
            entityId: hl7Message.id,
            userId: user.id,
            details: { messageType, sourceSystem },
        });
        return {
            messageId: hl7Message.id,
            messageType,
            status: 'PROCESSED',
            result,
        };
    }
    async processAdmitMessage(admitDto, user) {
        this.logger.log(`Processing HL7 admit message`);
        const { patientId, admissionDate, admittingDoctor, admissionType } = admitDto;
        let patient = await this.prisma.patient.findUnique({
            where: { id: patientId },
        });
        if (!patient) {
            patient = await this.prisma.patient.create({
                data: {
                    id: patientId,
                    tenantId: user.tenantId,
                },
            });
        }
        const admission = await this.prisma.patientAdmission.create({
            data: {
                patientId: patient.id,
                admissionDate: new Date(admissionDate),
                admittingDoctorId: admittingDoctor,
                admissionType,
                status: 'ADMITTED',
                admittedBy: user.id,
            },
        });
        return { admissionId: admission.id, patientId: patient.id };
    }
    async processDischargeMessage(dischargeDto, user) {
        this.logger.log(`Processing HL7 discharge message`);
        const { patientId, dischargeDate, dischargeDisposition } = dischargeDto;
        const admission = await this.prisma.patientAdmission.findFirst({
            where: {
                patientId,
                status: 'ADMITTED',
            },
            orderBy: { admissionDate: 'desc' },
        });
        if (!admission) {
            throw new common_1.BadRequestException('No active admission found for patient');
        }
        const dischargedAdmission = await this.prisma.patientAdmission.update({
            where: { id: admission.id },
            data: {
                dischargeDate: new Date(dischargeDate),
                dischargeDisposition,
                status: 'DISCHARGED',
                dischargedBy: user.id,
            },
        });
        return { admissionId: dischargedAdmission.id, patientId };
    }
    async processObservationMessage(observationDto, user) {
        this.logger.log(`Processing HL7 observation message`);
        const { patientId, observations } = observationDto;
        const observationRecords = [];
        for (const obs of observations) {
            const observation = await this.prisma.observation.create({
                data: {
                    patientId,
                    name: obs.name,
                    value: obs.value,
                    unit: obs.unit,
                    status: 'FINAL',
                    category: obs.category,
                    effectiveDateTime: new Date(obs.effectiveDateTime),
                },
            });
            observationRecords.push(observation);
        }
        return {
            patientId,
            observationCount: observationRecords.length,
            observations: observationRecords.map(o => o.id),
        };
    }
    async processOrderMessage(orderDto, user) {
        this.logger.log(`Processing HL7 order message`);
        const { patientId, orderType, orderDetails } = orderDto;
        const order = await this.prisma.labOrder.create({
            data: {
                patientId,
                orderType,
                orderDetails: JSON.stringify(orderDetails),
                status: 'ORDERED',
                orderedBy: user.id,
                orderedAt: new Date(),
            },
        });
        return { orderId: order.id, patientId };
    }
    async getHl7Message(messageId, user) {
        this.logger.log(`Getting HL7 message: ${messageId}`);
        const message = await this.prisma.hl7Message.findFirst({
            where: {
                id: messageId,
                tenantId: user.tenantId,
            },
        });
        if (!message) {
            throw new Error('HL7 message not found');
        }
        return {
            id: message.id,
            messageType: message.messageType,
            messageContent: message.messageContent,
            sourceSystem: message.sourceSystem,
            processedAt: message.processedAt,
            status: message.status,
        };
    }
    async syncData(messageType, syncFrom, syncTo, tenantId) {
        this.logger.log(`Syncing HL7 data: ${messageType} from ${syncFrom} to ${syncTo}`);
        const syncResult = {
            messageType,
            syncedFrom: syncFrom,
            syncedTo: syncTo,
            recordsProcessed: 0,
            errors: [],
        };
        await new Promise(resolve => setTimeout(resolve, 1000));
        return syncResult;
    }
    async getStatus(tenantId) {
        this.logger.log(`Getting HL7 status for tenant: ${tenantId}`);
        const config = await this.prisma.integrationConfiguration.findFirst({
            where: {
                tenantId,
                integrationType: 'HL7',
            },
        });
        return {
            integrationType: 'HL7',
            status: config?.isActive ? 'HEALTHY' : 'ERROR',
            lastSyncAt: config?.lastSyncAt,
            isActive: config?.isActive || false,
        };
    }
    async getStats(tenantId) {
        this.logger.log(`Getting HL7 stats for tenant: ${tenantId}`);
        const messageCount = await this.prisma.hl7Message.count({ where: { tenantId } });
        const processedCount = await this.prisma.hl7Message.count({
            where: {
                tenantId,
                status: 'PROCESSED',
            },
        });
        const failedCount = await this.prisma.hl7Message.count({
            where: {
                tenantId,
                status: 'FAILED',
            },
        });
        const messagesByType = await this.prisma.hl7Message.groupBy({
            by: ['messageType'],
            where: { tenantId },
            _count: { messageType: true },
        });
        return {
            totalMessages: messageCount,
            processedMessages: processedCount,
            failedMessages: failedCount,
            messagesByType,
        };
    }
    async retryOperation(log) {
        this.logger.log(`Retrying HL7 operation: ${log.id}`);
        const retryResult = {
            success: true,
            message: 'Operation retried successfully',
        };
        return retryResult;
    }
    parseHl7Message(messageContent) {
        this.logger.log(`Parsing HL7 message: ${messageContent.substring(0, 100)}...`);
        return {
            messageType: 'ADT_A01',
            patientId: 'PAT001',
            patientName: 'John Doe',
        };
    }
};
exports.Hl7Service = Hl7Service;
exports.Hl7Service = Hl7Service = Hl7Service_1 = __decorate([
    (0, common_1.Injectable)(),
    __metadata("design:paramtypes", [prisma_service_1.PrismaService,
        audit_service_1.AuditService])
], Hl7Service);
